### ISSUE TITLE
Randomization seed module for PhotonDust

### ISSUE THREAD

**Issue by TheBasementNerd** [Yellow-Dog-Man/Resonite-Issues#3425]
### Is your feature request related to a problem? Please describe.

Particles spawn with local randomization and cannot be synchronized globally

### Describe the solution you'd like

A module that can be added to a PhotoDust particle system where the player can enter a seed for the RNG that allows the randomization of the system to be synchronized across all users, so rotations, positions, and value ranges upon a particle's spawn will be the same for all users 

### Describe alternatives you've considered

There is no alternative 

### Requesters

TheBasementNerd

**Comment by Frooxius** [comment]
What you propose wouldn't really solve the problem and give what you want. This problem is A LOT more complicated than you might even realize.

Making sure the initial seed is the same for every single user is the trivial part - the really difficult one is making sure that the system does not diverge over the course of its simulation. Consider all things that can happen to a particle system:
- Emitters must exist EXACTLY the same amount of particles at exact same points in time - if even one emitter emits one extra or less particles at particular frame (e.g. due to users running at different framerates), now the random sequence is off and it had diverged
- Users needs to all simulate them at EXACT same rate - this on itself can be really tricky, since each user joins at different speeds, have hiccups that happen at different times. For this to work, all particle systems need to run in perfect lock-step
- Any parameters that are changed or driven - these must change EXACTLY to the same values for each user at the EXACT same points in simulation time - anything else can cause divergence
- The seed only works for the initial state - so it'll only work when all users are present. Once other users join in, the system is already been simulating and they don't have that state
- Multi-threading - any multi-threaded parts can potentially lead to divergence, because they're stochastic
- HW differences - floating point implementation differences can cause particle systems to diverge from each other even if everything else is exactly the same

In short... this is a VERY complicated problem to solve, with lots of potential overhead and not something we plan on solving through PhotonDust directly - this is not one of its design goals.

Because of that, we'll not be implementing this.

**Comment by TheBasementNerd** [comment]
While I understand the complication that would go into making the simulation run the exact same for every user at the same time, my initial request was specifically about the initial state value of particles emitted being the same for each user, I had no intention of suggesting we have the entire simulation locked into sync

**Comment by Frooxius** [comment]
I understand that, but you're also asking for it in order for the particle systems to be synchronized across all users - this will NOT do that.

We're generally not going to add features that are a flawed "solution" to a problem (as in, they appear like a solution, but are not actually solution at all).

**Comment by TheBasementNerd** [comment]
No, I'm not asking for it so that the simulation is synchronized, I think you're assuming that's my intent, but I never suggested that. I'm simply asking for a feature that allows the seed to result in the same values _at the time of emission_, a time that I imagine comes from something global such as world time. If the simulation was synchronized, driven values that are localized differently for players wouldn't work

**Comment by Frooxius** [comment]
That's what you're specifically asking for in your post though:

> allows the randomization of the system to be synchronized across all users, so rotations, positions, and value ranges upon a particle's spawn will be the same for all users

I understand that you want this to happen at emission time. But what I'm saying is that this will diverge. 

You cannot disentangle the emission from simulation - they are the same process and they both potentially use random values - e.g. if the emission differs slightly between users, now the randomization is desynced. 

With that you're asking for, it might be same for a little bit, but sooner or later even the emission itself will diverge, due to variances in the emission process and other parts of simulation that use random values.

**Comment by DoubleStyx** [comment]
In other words, as long as the starting seed is the same, the system will begin in the same configuration. While it will naturally diverge over time, this property can still be leveraged for short-term consistency across users. This wouldn't attempt to present as a true syncing mode; this would just be providing a seed value as a syncvar.

**Comment by TheBasementNerd** [comment]
> In other words, as long as the starting seed is the same, the system will begin in the same configuration. While it will naturally diverge over time, this property can still be leveraged for short-term consistency across users, not for long-running sims. This wouldn't attempt to present as a true syncing mode; this would just be providing a seed value as a syncvar.

Correct. I am not looking to have the natural randomization and divergence be synchronized, just the initial values of each particle upon spawn at a certain time



> That's what you're specifically asking for in your post though:
> 
> > allows the randomization of the system to be synchronized across all users, so rotations, positions, and value ranges upon a particle's spawn will be the same for all users
> 
> I understand that you want this to happen at emission time. But what I'm saying is that this will diverge.

Yes I recognize it will change and diverge over time or after spawn. I am referring to the RNG randomization, not the natural randomization that comes from it being on different machines

**Comment by Frooxius** [comment]
> Yes I recognize it will change and diverge over time or after spawn. I am referring to the RNG randomization, not the natural randomization that comes from it being on different machines

Yes, I was also talking about RNG randomization - but also highlighting the fact that this is tied into the natural randomization - The natural randomization is what randomizes the RNG. What I am trying to point out is that you **cannot** separate the two - they are intertwined. 

The simulation influences when and how the RNG gets sampled. So if there are differences in the simulation, it will change how the RNG is sampled (as well as its state) resulting in different state of each user. 

It can potentially diverge very fast, not enough to really be useful.

As a principle, we do not offer features like this, because they provide "false" synchronization - they might look synchronized at first, but it can be hard to tell when, how and how much they diverge. We don't want people to rely on systems for synchronization that are fundamentally not reliable for that.

**Comment by DoubleStyx** [comment]
Ah, I see. The system itself would have to essentially not be random at that point, so a seed value wouldn't work here. Instead, this would have to be done through the initial configuration itself, which sort of defeats the point of RNG in most cases. Unless rather than using a true cryptographic seed, this would be more of a smoothly time-varying generator seed, similar to perlin noise.

**Comment by TheBasementNerd** [comment]
So each particle samples the RNG function values in some way that isn't tied just to current time at emission? That's just what I imagine an RNG function in this use case does, at least

**Comment by DoubleStyx** [comment]
RNG in most cases tries to generate statistically random samples based on data/entropy coming from your system. It's usually handled through something that's random-looking on the surface but isn't random enough for cryptographic use. So overall, the seed in most cases does not really give deterministic time-varying output, though this varies on the implementation.

**Comment by Frooxius** [comment]
@TheBasementNerd They potentially can. But the main thing that really confuses it here is the emission itself. The emission is part of the simulation process - the emitters will accumulate particles to emit. This accumulation will differ depending on the framerate of the user, meaning each frame, each user emits a different number of particles from the emitters, in potentially different orders - the system doesn't guarantee that this is deterministic. It's also heavily multi-threaded in parts (and there's more planned), which also removes guarantees of determinism.

In order for RNG to produce the same results, you need the system to be deterministic, but PhotonDust is not designed to be that, so it doesn't mesh with this at all - like I mentioned earlier, it's not one of its design goals.

**Comment by TheBasementNerd** [comment]
So the misunderstanding here was that I did not recognize how the RNG was built under the hood. I presumed it to simply be sampling the RNG values based on time, and did not realize that there may be other systems that affect it even before emission. I've never worked with randomization functions in any other capacity, but with how you describe its workings, I understand why it wouldn't work

**Comment by Frooxius** [comment]
Possibly. Generally the way RNG works, when given an initial seed value is that it will give you the same sequence of pseudo-random values. Time doesn't matter, only the order you sample it at does. 

Which means if you want to get exact same behavior, it needs to be sampled in exact same order for every single part of the system, so they each get the exact same sequence.

In order for that to work, the system must be designed to be fully deterministic, even with other varying conditions, which makes this really difficult.
